File: DSCI 553/Homework 3/task2predict.py
import json
from pyspark import SparkContext
import sys
import math

sc = SparkContext("local[*]", "homework3_task2predict")

#python task2predict.py dataset/test_review.json task2model.model task2.predict 
args = sys.argv

input_file = args[1]
model_path = args[2]
output_path = args[3]

model = sc.textFile(model_path).map(json.loads) \
    .map( lambda x: (x['type'], x['id'], x['value']) )


userProfiles = model.filter( lambda x: x[0] == 'User' ) \
    .map( lambda x: (x[1], x[2]) ) \
    .collectAsMap()

docProfiles = model.filter( lambda x: x[0] == 'Document' ) \
    .map( lambda x: (x[1], x[2]) ) \
    .collectAsMap()

def cosine_similarity(bid, uid, userProfiles, docProfiles):
    if uid not in userProfiles or bid not in docProfiles: return 0
    uv, bv = set(userProfiles[uid]), set(docProfiles[bid])
    return len(uv.intersection(bv)) / (math.sqrt(len(uv)) * math.sqrt(len(bv)))

predicted = sc.textFile(input_file).map( json.loads ) \
    .map( lambda x: (x['business_id'], x['user_id']) ) \
    .map( lambda x: (x[0], x[1], cosine_similarity(x[0], x[1], userProfiles, docProfiles) ) ) \
    .filter(lambda x: x[2] >= 0.01) \
    .map( lambda x: "{"+f'"user_id":"{x[1]}", "business_id":"{x[0]}", "sim":{x[2]}'+"}" ) \
    .collect()

with open(output_path, "w") as out:
    out.write("\n".join(predicted))

File: DSCI 553/Homework 3/task2predict.scala
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession
import org.json4s.jackson.JsonMethods

import java.io.PrintWriter
import scala.collection.mutable.ListBuffer
import scala.collection.{immutable, mutable}

object task2predict {

  //spark-submit --driver-memory 4g --executor-memory 4g --class task2predict hw3.jar $ASNLIB/publicdata/test_review.json task2.scala.model task2.scala.predict
  def cosineSimilarity( bid: String, uid: String,
                        userProfiles:  scala.collection.Map[String, Set[Int]],
                        businessProfiles: scala.collection.Map[String, Set[Int]]): Double ={

    if ( (userProfiles.getOrElse(uid, null) == null) | (businessProfiles.getOrElse(bid, null) == null)) {0.0}
    else{
      val uv = userProfiles(uid)
      val bv = businessProfiles(bid)
      (1.0*uv.intersect(bv).size) / (math.sqrt(uv.size) * math.sqrt(bv.size))
    }
  }


  def main(args: Array[String]) {
    Logger.getLogger("org").setLevel(Level.ERROR)

    val ss = SparkSession
      .builder()
      .appName("task2train")
      .config("spark.master", "local[*]")
      .getOrCreate()

    val input_file = args(0)
    val model_file = args(1)
    val output_file = args(2)

    val sc = ss.sparkContext

    val model = sc.textFile(model_file).map( x => JsonMethods.parse(x).values.asInstanceOf[Map[String, Any]] )
      .map( x => ( (x.getOrElse("type", null).asInstanceOf[String],
                    x.getOrElse("id", null).asInstanceOf[String],
                    x.getOrElse("value", null).asInstanceOf[List[Int]].toSet)))

    val userProfiles = model.filter( x => x._1 == "User" )
      .map(x=> (x._2, x._3))
      .collectAsMap()

    val docProfiles = model.filter( x => x._1 == "Document" )
      .map( x => (x._2, x._3) )
      .collectAsMap()

    val predicted = sc.textFile(input_file)
      .map( x => JsonMethods.parse(x).values.asInstanceOf[Map[String, Any]] )
      .map( x => ( x.getOrElse("business_id", null).asInstanceOf[String],
                    x.getOrElse("user_id", null).asInstanceOf[String]))
      .map( x=> ( x._1, x._2, cosineSimilarity(x._1, x._2, userProfiles, docProfiles) ) )
      .filter( x => x._3 >= 0.01 )
      .map( x => "{"+"\"user_id\": \""+x._2+"\", \"business_id\": \""+x._1+"\", \"sim\": "+x._3.toString+"}" )
      .collect()

    new PrintWriter(output_file) { write(predicted.mkString("\n")); close }
  }

}
